pkg/controllers/management/k3supgrade/k3supgrade_handler.go

```diff
@@ -0,0 +1,174 @@
+	utils2 "github.com/rancher/rancher/pkg/app/utils"
+	"github.com/rancher/rancher/pkg/catalog/utils"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rancher/pkg/project"
+	"github.com/rancher/rancher/pkg/ref"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	v32 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (h *handler) onClusterChange(key string, cluster *v3.Cluster) (*v3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	// only applies to k3s clusters
+	if cluster.Status.Driver != v3.ClusterDriverK3s {
+		return cluster, nil
+	}
+
+	// setup default k3sconfig
+	if cluster.Spec.K3sConfig == nil {
+		version := *cluster.Status.Version
+		cluster.Spec.K3sConfig = &v3.K3sConfig{
+			Version: version.String(),
+		}
+	}
+
+	if cluster.Spec.K3sConfig.Version == "" {
+		cluster.Spec.K3sConfig.Version = cluster.Status.Version.String()
+	}
+
+	isNewer, err := IsNewerVersion(cluster.Status.Version.GitVersion, cluster.Spec.K3sConfig.Version)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !isNewer {
+		// cannot upgrade- version has not changed or proposed version is not newer than current version
+		return cluster, nil
+	}
+
+	// create or update k3supgradecontroller if necessary
+	if err = h.deployK3sUpgradeController(cluster.Name); err != nil {
+		return cluster, err
+	}
+
+	// deploy plans into downstream cluster
+	if err = h.deployPlans(cluster); err != nil {
+		return cluster, err
+	}
+
+	return cluster, nil
+}
+
+// deployK3sUpgradeController creates a rancher k3s upgrader controller if one does not exist.
+// Updates k3s upgrader controller if one exists and is not the newest available version.
+func (h *handler) deployK3sUpgradeController(clusterName string) error {
+	userCtx, err := h.manager.UserContext(clusterName)
+	if err != nil {
+		return err
+	}
+
+	projectLister := userCtx.Management.Management.Projects("").Controller().Lister()
+	systemProject, err := project.GetSystemProject(clusterName, projectLister)
+	if err != nil {
+		return err
+	}
+
+	templateID := k3sUpgraderCatalogName
+	template, err := h.templateLister.Get(namespace.GlobalNamespace, templateID)
+	if err != nil {
+		return err
+	}
+
+	latestTemplateVersion, err := utils.LatestAvailableTemplateVersion(template)
+	if err != nil {
+		return err
+	}
+
+	creator, err := h.systemAccountManager.GetSystemUser(clusterName)
+	if err != nil {
+		return err
+	}
+	systemProjectID := ref.Ref(systemProject)
+	_, systemProjectName := ref.Parse(systemProjectID)
+
+	nsClient := userCtx.Core.Namespaces("")
+	appProjectName, err := utils2.EnsureAppProjectName(nsClient, systemProjectName, clusterName, systemUpgradeNS, creator.Name)
+	if err != nil {
+		return err
+	}
+
+	appLister := userCtx.Management.Project.Apps("").Controller().Lister()
+	appClient := userCtx.Management.Project.Apps("")
```

Comment:
Why don't these live on handler?

Reply:
because the handler on the management cluster but the app is being deployed in the downstream cluster, which will be cluster specific.

Reply:
That doesn't make sense. App is a rancher CRD and this pathing is calling management, which gets back to the management context. This is not doing anything with the downstream cluster.